### PR TITLE
Multi variable declarations

### PR DESCRIPTION
**Pull request by chriseth**
Closes https://github.com/ethereum/solidity/issues/3520
Closes https://github.com/ethereum/solidity/issues/3669

Depends on #4047.

TODO:

 - [x] simplify changes in Parser
 - [x] some end to end tests
 - [x] documentation

### CODE REVIEW

**Comment by chriseth** [comment]
Split this up into multiple steps, should be easy to review.

File: docs/control-structures.rst [hunk 1]
```diff
@@ -272,7 +272,10 @@ Assignment
 Destructuring Assignments and Returning Multiple Values
 -------------------------------------------------------
 
-Solidity internally allows tuple types, i.e. a list of objects of potentially different types whose size is a constant at compile-time. Those tuples can be used to return multiple values at the same time and also assign them to multiple variables (or LValues in general) at the same time::
+Solidity internally allows tuple types, i.e. a list of objects of potentially different types whose size is a constant at compile-time. Those tuples can be used to return multiple values at the same time.
+These can then either be assigned to newly declared variables or to pre-existing variables (or LValues in general):
+
+::
 
     pragma solidity ^0.4.16;
```
Comment (axic) [inline docs/control-structures.rst]: Needs a version bump.

File: docs/frequently-asked-questions.rst [hunk 2]
```diff
@@ -219,7 +219,7 @@ If you do not want to throw, you can return a pair::
         }
```
Comment (axic) [inline docs/frequently-asked-questions.rst:219]: Needs a version bump above too.

File: docs/solidity-by-example.rst [hunk 3]
```diff
@@ -467,8 +467,8 @@ high or low invalid bids.
```
Comment (axic) [inline docs/solidity-by-example.rst:467]: Same here.

**Comment by axic** [comment]
Can the refactorings split out?

**Comment by chriseth** [comment]
Split refactorings out into https://github.com/ethereum/solidity/pull/4047

**Comment by chriseth** [comment]
Rebased on top of  #4047 and updated documentation versions.

File: docs/control-structures.rst (hunk 1 above)
Comment (axic) [reply docs/control-structures.rst]: This was missed.

**Comment by axic** [comment]
Please rebase.

**Comment by chriseth** [comment]
Rebased.

File: docs/control-structures.rst (hunk 1 above)
Comment (axic) [reply docs/control-structures.rst]: @chriseth this version bump is still missing :)

Comment (chriseth) [reply docs/control-structures.rst]: argh, sorry.

Comment (chriseth) [reply docs/control-structures.rst]: Fixed.

File: Changelog.md [hunk 4]
```diff
@@ -4,6 +4,7 @@ Features:
  * Build System: Update internal dependency of jsoncpp to 1.8.4, which introduces more strictness and reduces memory usage.
  * Code Generator: Use native shift instructions on target Constantinople.
  * Gas Estimator: Only explore paths with higher gas costs. This reduces accuracy but greatly improves the speed of gas estimation.
+ * General: Allow multiple variables to be declared using ``(uint a, uint b) = ...``.
```
Comment (axic) [inline Changelog.md]: I think this is misleading. It doesn't allow multiple variable declarations on their own, only declaration with assignment.

File: test/libsolidity/syntaxTests/multiVariableDeclaration/multiVariableDeclarationComplex.sol [hunk 5]
```diff
@@ -0,0 +1,9 @@
+contract D {
+  struct S { uint a; uint b; }
+}
+contract C {
+  function f() internal returns (uint, uint, uint, D.S[20] storage, uint) {
+    (,,,D.S[10*2] storage x,) = f();
```
Comment (axic) [inline test/libsolidity/syntaxTests/multiVariableDeclaration/multiVariableDeclarationComplex.sol:6]: These will be kind of hard to spot and read :)

File: Changelog.md (hunk 4 above)
Comment (chriseth) [reply Changelog.md]: That's why I put the example there. Any suggestion for a better text?

**Comment by chriseth** [comment]
@axic can this be merged?

File: Changelog.md (hunk 4 above)
Comment (axic) [reply Changelog.md]: How about 

> Allow multiple typed variables to be declared as part of a tuple assignment, e.g. ``(uint a, uint b) = ...``

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
Rebased this and it caused failures:
```
ASSERTION FAILURE:
- file   : SyntaxTest.cpp
- line   : 274
- message: Test expectation mismatch.
Expected result:
  Success
Obtained result:
  Warning: (163-169): This variable is of storage pointer type and might be returned without assignment. This can cause storage corruption. Assign the variable (potentially from itself) to remove this warning.
```

**Comment by axic** [comment]
I think this and #4138 may have cross dependencies, so one of them must be rerun after a merge.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Fixed. Should be merged first.

**Comment by Pzixel** [comment]
How is it expected to work?

I have a function

```
function getRequestByIndex(uint64) public view returns (bytes30, uint64, uint64, string, uint64, uint64[], uint64[], uint64, uint64, bytes16, string) {
	return (0, 0, 0, "", 0, new uint64[](0), new uint64[](0), 0, 0, 0, "");
}
```

I call it:

```
(bytes30 serviceNumber, uint64 requestDate, uint64 declarantType, string declarantName, uint64 fairId, uint64[] assortment, uint64[] periods, uint64 district, uint64 region, bytes16 userId, string details) = season.getRequestByIndex(index);
```

Result:

```
Error	G8E0852CE	Type uint64[] memory is not implicitly convertible to expected type uint64[] storage pointer.
Error	MSB3073 	The command "solc "Contract.sol" -o "obj/Debug" --bin --abi --overwrite" exited with code 1.	
Error	G00820F71	Type string memory is not implicitly convertible to expected type string storage pointer.	
Error	G00820F71	Type string memory is not implicitly convertible to expected type string storage pointer.	
Error	G8E0852CE	Type uint64[] memory is not implicitly convertible to expected type uint64[] storage pointer.
```

---

On the other hand I can write

```
var (serviceNumber, requestDate, declarantType, declarantName, fairId, assortment, periods, district, region, userId, details) = season.getRequestByIndex(index);
```

That indeed work, but compiler places lots of warnings about `unused variable`. I also tried to wirte:

```
bytes30 serviceNumber;
uint64 requestDate;
uint64 declarantType;
string memory declarantName;
uint64 fairId;
uint64[] memory assortment;
uint64[] memory periods;
uint64 district;
uint64 region;
bytes16 userId;
(serviceNumber, requestDate, declarantType, declarantName, fairId, assortment, periods, district, region, userId, ) = season.getRequestByIndex(index);
```

Which complains that `stack too deep`.

**Comment by chriseth** [comment]
You have to use `(..., string memory declarantName, ...)`.

**Comment by Pzixel** [comment]
Thank you, I managed to compile it with following signature
`(bytes30 serviceNumber, uint64 requestDate, Types.DeclarantType declarantType, string memory declarantName, uint64 fairId, Types.Speciality[] memory assortment, uint64 district, uint64 region, string memory details, uint64[] memory periods, bytes16 userId) = season.getRequestByIndex(index);`

**Pull request merged**